[cocoa] Add quirk for billpaysite.com

sysrqb opened the pull request:
#### b0f069e44cd98013c9db57ec2e031111bbcead45
<pre>
[cocoa] Add quirk for billpaysite.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=284503">https://bugs.webkit.org/show_bug.cgi?id=284503</a>
<a href="https://rdar.apple.com/141328971">rdar://141328971</a>

Reviewed by Wenson Hsieh.

As a third party, billpaysite.com relies on access to cookies. These cookies
don&apos;t need to be first party cookies, because a user typically doesn&apos;t interact
with billpaysite.com as a first party. Unfortunately, the site has not adopted
CHIPS yet, either. As a workaround, we can leverage CHIPS and opt the site into
partitioned cookies when it is loaded as a third party.

On Cocoa platforms, we use the _cookieTransform callback where we modify the
cookie&apos;s partition key before it is saved in the cookie store.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsPartitionedCookies):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::setCookiePartition):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::setCookieTransformForFirstPartyRequest):
(WebKit::NetworkTaskCocoa::setCookieTransformForThirdPartyRequest):
(WebKit::NetworkTaskCocoa::setCookieTransform):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkTaskCocoa::applyCookiePolicyForThirdPartyCloaking): Deleted.

Canonical link: <a href="https://commits.webkit.org/288096@main">https://commits.webkit.org/288096@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7673603f47f9430d8a4b64c478d92d08a8a22e94

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81873 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1399 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35827 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86427 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32869 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83979 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9222 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63861 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21586 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84943 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1055 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74507 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44146 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/954 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28684 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31322 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29293 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87857 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9113 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6494 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72217 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9298 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70328 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71439 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15532 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14455 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12684 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9064 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14596 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8905 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12428 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10713 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -102,6 +103,16 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+{
+    auto *partitionedCookies = [NSMutableArray arrayWithCapacity:cookies.count];
+    for (NSHTTPCookie *cookie in cookies)
+        [partitionedCookies addObject:WebCore::NetworkStorageSession::setCookiePartition(cookie, partition)];
```
It looks like `+cookieWithProperties:` returns a nullable `NSHTTPCookie *`, so we should probably null check this to be safe (instead of crashing).

sysrqb replied:
Seems like we need to do a better job of this, in general.

whsieh commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1885,6 +1886,13 @@ bool Quirks::needsDesktopUserAgent(const URL& url)
+{
+    if (request.isTopSite())
+        return false;
+    return request.url().protocolIs("https"_s) && request.url().host().endsWith(".billpaysite.com"_s);
```
Does this need to be limited to `https`? I think we generally use `URL::protocolIsInHTTPFamily()` for this (which just looks up a cached bit).

sysrqb replied:
I suppose not. Using a bill payment site with http would make me uncomfortable, but you're right that this quirk doesn't need to be limited to https.

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -145,6 +179,11 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+#if HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)
+        , needsQuirk
+        , isRequestThirdParty = request.isThirdParty()
+        , cookiePartition = crossThreadCopy(cookiePartition)
```
Why not only capture `cookiePartition` and set `cookiePartition` to null if we don't need quirk (i.e. `!needsQuirk`)?

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -127,15 +141,35 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+    }
+
+    auto cookiePartition = networkStorageSession->cookiePartitionIdentifier(request);
+#else
```
Is this correct? 
Before this change, if `shouldApplyCookiePolicyForThirdPartyCloaking` is true and `networkStorageSession->isOptInCookiePartitioningEnabled()` is false, we will set `_cookieTransformCallback`, if request is first-party; with this change it seems we don't set it (due to the early return at line 160).

It might be clearer to not mix these two cases. My understanding is one will only be applied to first-party request and the other is only applied to third-party request, so maybe we could have`setCookieTransformForFirstPartyRequest` and `setCookieTransformForThirdPartyRequest`, and ensure only one setting will succeed.

sysrqb replied:
Good catch, that check should be `needsQuirk && !networkStorageSession->isOptInCookiePartitioningEnabled()`. We could split these into two functions, but I was concerned about them conflicting in the future. I agree they should be mutually exclusive right now.

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -346,6 +346,18 @@ - (CFURLRef)_cfurl;
+NSHTTPCookie *NetworkStorageSession::setCookiePartition(NSHTTPCookie *cookie, NSString* partitionKey)
+{
+    if (cookie._storagePartition)
+        return cookie;
```
Should we assert `_storagePartition` is the same as `partitionKey`?

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -346,6 +346,18 @@ - (CFURLRef)_cfurl;
 
+#if HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)
+NSHTTPCookie *NetworkStorageSession::setCookiePartition(NSHTTPCookie *cookie, NSString* partitionKey)
+{
```
Should we return early if `partitionKey` is nil, or assert it's not empty?

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -102,6 +103,19 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+{
+    auto *partitionedCookies = [NSMutableArray arrayWithCapacity:cookies.count];
+    for (NSHTTPCookie *cookie in cookies) {
+        NSHTTPCookie *partitionedCookie = WebCore::NetworkStorageSession::setCookiePartition(cookie, partition);
```
`RetainPtr partitionedCookie`

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -102,6 +103,19 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+#if HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)
+static NSArray<NSHTTPCookie *> *cookiesBySettingPartition(NSArray<NSHTTPCookie *> *cookies, NSString* partition)
+{
+    auto *partitionedCookies = [NSMutableArray arrayWithCapacity:cookies.count];
```
`RetainPtr<NSMutableArray`

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -102,6 +103,19 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+    for (NSHTTPCookie *cookie in cookies) {
+        NSHTTPCookie *partitionedCookie = WebCore::NetworkStorageSession::setCookiePartition(cookie, partition);
+        if (partitionedCookie)
+            [partitionedCookies addObject:partitionedCookie];
```
It seems partitioned cookie cannot be nil after `setCookiePartition` (unless `cookie` is nil, and if `cookie` can be nil we probably need null check and early return in `setCookiePartition`)

Commit:
[cocoa] Add quirk for billpaysite.com
https://bugs.webkit.org/show_bug.cgi?id=284503
rdar://141328971

Reviewed by Wenson Hsieh.

As a third party, billpaysite.com relies on access to cookies. These cookies
don't need to be first party cookies, because a user typically doesn't interact
with billpaysite.com as a first party. Unfortunately, the site has not adopted
CHIPS yet, either. As a workaround, we can leverage CHIPS and opt the site into
partitioned cookies when it is loaded as a third party.

On Cocoa platforms, we use the _cookieTransform callback where we modify the
cookie's partition key before it is saved in the cookie store.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsPartitionedCookies):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::setCookiePartition):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::setCookieTransformForFirstPartyRequest):
(WebKit::NetworkTaskCocoa::setCookieTransformForThirdPartyRequest):
(WebKit::NetworkTaskCocoa::setCookieTransform):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkTaskCocoa::applyCookiePolicyForThirdPartyCloaking): Deleted.

Canonical link: https://commits.webkit.org/288096@main
